src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1023 @@
+# ${build-info}
+#
+#
+# This module implements a rule-based editor that is used to modify the content
+# of an existing file without taking care of the whole file. Each rule
+# driving the edition process is applied to all matching lines. The input for
+# updating the file is the Quattor configuration and conditions can be defined
+# based on the contents of this configuration.
+#
+# This module is a subclass of the CAF FileEditor: it extends the base methods of 
+# the CAF FileEditor. In addition to the constructor it has only one public method.
+# The methods provided in this module can be used at the same time as the
+# base methods of the FileEditor.
+#
+#
+#######################################################################
+
+package CAF::RuleBasedEditor;
+
+use strict;
+use warnings;
+use vars qw(@ISA $EC);
+$EC=LC::Exception::Context->new->will_store_all;
+
+use parent qw(CAF::FileEditor Exporter);
+
+use EDG::WP4::CCM::Element;
+
+use Readonly;
+
+use Encode qw(encode_utf8);
+
+local(*DTA);
+
+# Constants from FileEditor
+use CAF::FileEditor qw(BEGINNING_OF_FILE ENDING_OF_FILE);
+
+
+#########################################################
+# Constants use to format lines in configuration files. #
+# These constants are exported.                         #
+#########################################################
+
+#  LINE_FORMAT_xxx: general syntax of the line (key/val format)
+# LINE_FORMAT_SH_VAR:         key=val (e.g. SH shell family)
+# LINE_FORMAT_ENV_VAR:        export key=val (e.g. SH shell family)
+# LINE_FORMAT_KEY_VAL:        key val (e.g. Xrootd, Apache)
+# LINE_FORMAT_KEY_VAL_SETENV: setenv key val  (used by Xrootd in particular)
+# LINE_FORMAT_KEY_VAL_SET:    set key val  (used by Xrootd in particular)
+use enum qw(LINE_FORMAT_SH_VAR=1
+            LINE_FORMAT_ENV_VAR
+            LINE_FORMAT_KEY_VAL
+            LINE_FORMAT_KEY_VAL_SETENV
+            LINE_FORMAT_KEY_VAL_SET
+           );
+
+# LINE_VALUE_xxx: how to interpret the configuration value
+# LINE_VALUE_INSTANCE_PARAMS is specific to Xrootd.
+use enum qw(LINE_VALUE_AS_IS
+            LINE_VALUE_BOOLEAN
+            LINE_VALUE_ARRAY
+            LINE_VALUE_HASH_KEYS
+            LINE_VALUE_STRING_HASH
+            LINE_VALUE_INSTANCE_PARAMS
+           );
+
+# LINE_VALUE_OPT_xxx: options for rendering the value
+# (mainly apply to lists and dictionnaries)
+use enum qw(BITMASK: LINE_VALUE_OPT_SINGLE
+                     LINE_VALUE_OPT_UNIQUE
+                     LINE_VALUE_OPT_SORTED
+           );
+
+# Internal constants
+Readonly my $LINE_FORMAT_DEFAULT => LINE_FORMAT_SH_VAR;
+Readonly my $LINE_QUATTOR_COMMENT => "\t\t# Line generated by Quattor";
+Readonly my $LINE_OPT_DEF_REMOVE_IF_UNDEF => 0;
+Readonly my $LINE_OPT_DEF_ALWAYS_RULES_ONLY => 0;
+Readonly my $RULE_CONDITION_ALWAYS => 'ALWAYS';
+Readonly my $RULE_OPTION_SET_GLOBAL => 'GLOBAL';
+
+
+# Export constants used to build rules
+Readonly my @RULE_CONSTANTS => qw(LINE_FORMAT_SH_VAR
```

Comment:
I have to check. I remember finding one when I started this work in the framework of `ncm-dpmlfc`. But this was relying on a `BEGIN` block and that was failing because it was overriding the `BEGIN` block from `ncm-dpmlfc`, due to the fact that this was not a separate class. May work now... I'll have a look.

The risk of bad mistakes is almost 0 as if you remove a constant and forget to remove it from this list, PErl will complain and if you add a constant and forget to export it, it is not usable and users will complain!!